src/librustc/hir/lowering.rs

```diff
@@ -3021,6 +3016,11 @@ impl<'a> LoweringContext<'a> {
                 });
                 let impl_trait_return_allow = !self.is_in_trait_impl;
 
+                let asyncness = match sig.header.asyncness {
```

Comment:
I see this copy-pasted a few places-- can you break this out into a `return_impl_trait_id` method that returns an `Option<NodeId>`?